### PR TITLE
fixes #3048 - Non Admin user fails to override parameters when creating new host

### PR DESCRIPTION
**Pull request by isratrade**

### CODE REVIEW

File: app/models/lookup_value.rb [hunk 1]
```diff
@@ -61,11 +61,13 @@ def enforce_permissions operation
     return true if operation == "edit" and new_record?
     allowed = case match
       when /^fqdn=(.*)/
-        # check if current fqdn is in our allowed list
-        Host.my_hosts.where(:name => $1).exists?
+        # check if current fqdn is in our allowed list OR if it's a new host
+        # this will only be false if host exists and does not belong to you
+        !(Host.find_by_name($1) && Host.my_hosts.where(:name => $1).empty?)
```
Comment (isratrade) [inline app/models/lookup_value.rb]: @GregSutcliffe, does my logic make sense?

Comment (GregSutcliffe) [inline app/models/lookup_value.rb]: @isratrade I think so - I can test it later on, I still have my vm where I replicated the problem.

**Comment by GregSutcliffe** [comment]
Tested, fixes the issue. the new test seems sensible. :+1:

**Comment by domcleal** [comment]
I'm wondering about a different approach that wouldn't allow a user to create a lookup_value for any host that doesn't already exist (imagine if a user later created a host with that FQDN).

Currently the host / host group is responsible for creating the nested lookup values via the `lookup_values_attributes=` method in the HostCommon concern, so we could pass a reference to the host or flag into the new lookup value to indicate this particular instance is permitted.

e.g. an `attr_accessor :host` on LookupValue, pass in `self` to `host=` in `lookup_values_attributes=`, check `@host` in enforce_permissions and that $1 matches `@host.name`.

**Comment by isratrade** [comment]
@domcleal, if the approach is to **not** allow lookup_values created on host creation (since the host doesn't exist yet), then lookup_value.rb would stay the same and I would just edit the view to hide the "Override" button on the Parameters tab.

**Comment by domcleal** [comment]
Ah, that was a poor first sentence in my comment, sorry.

What I meant to say was: an approach that only allows the user to create a lookup_value for existing hosts, or a host that's in the process of creation.  What I'd like to disallow is a user with smart variable access being able to create a matcher for a host that doesn't exist (as it may do later and could bypass security).

See the rest of my reply for an idea about how it could be implemented.  What do you think?

**Comment by isratrade** [comment]
@domcleal, when a user creates a host, there is no custom matcher.  The override button automatically enters fqdn=host.name (I think), so there is no option for a host that doesn't exist (other than the host that he is creating).

**Comment by domcleal** [comment]
@isratrade a user could add a matcher through the Puppet Classes interface though, and the same enforce_permission method will be used to determine if they're allowed to add it there.

**Comment by isratrade** [comment]
@domcleal, I see.  I can't think how lookup_value.rb will allow a non-admin creating lookup_values on a new host (that doesn't yet exist) but not allowing lookup_values on a host that doesn't exist from the puppetclasses interface. It seem's we have have to add a check in the controller and not at the model level.

**Comment by domcleal** [comment]
@isratrade I think we could do it in `lookup_values_attributes=` in app/models/concerns/host_common.rb, as that's responsible for creating LookupValue records.  If we passed in a reference to self to the new lookup value, we could use this in enforce_permission?

**Comment by isratrade** [comment]
@domcleal, Non-admin users currently can't create a lookup_value for a host that doesn't exist, but admins can.  I assume this is what you want to prevent.

**Comment by isratrade** [comment]
@domcleal, lookup_values_attributes= in HostCommon uses def lookup_value_match in managed.rb and hostgroup.rb so these matches must exist as real hosts or hostgroups.

**Comment by isratrade** [comment]
@domcleal, re-submitted.  It's much more readable this way.  Good call on using attr_accessor on LookupValue.
I entered a separate issue #3206 - Prevent any user including admin from creating lookup_value with fdqn= to host that doesn't exist

**Comment by isratrade** [comment]
@GregSutcliffe, can you review this and see if it works in your case.  The test I added hopefully covers it.

**Comment by domcleal** [comment]
@isratrade I'm not really worried about the admin being able to do this, but the previous fix would have allowed non-admin users to bypass the old security mechanism.  I like this, thanks for the update.

Looks good, but I'll give it a test later unless @GregSutcliffe gets to it first.

**Comment by domcleal** [comment]
Thanks for the patch @isratrade, merged as 06909647b for Foreman 1.3.0-RC4.

**Comment by lzap** [comment]
Note for myself: This patch has wrong issue number, the correct one is:

http://projects.theforeman.org/issues/3076

**Comment by isratrade** [comment]
@lzap, woops, sorry about that

**Pull request closed without merging**